ceno_zkvm/src/instructions/riscv/mul.rs

```diff
@@ -246,21 +278,27 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for MulhInstructionBas
 
         // 2. Verify that the product of signed inputs `rs1` and `rs2` is equal to
         //    the result of interpreting `rd` as the high limb of a 2s complement
-        //    value with some 32-bit low limb
-
-        let prod_low = UInt::new(|| "prod_low", circuit_builder)?;
-        circuit_builder.require_equal(
-            || "validate_prod_high_limb",
-            rs1_val * rs2_val,
-            rd_val * (1u64 << 32) + prod_low.value(),
-        )?;
+        //    value
+        if I::INST_KIND == InsnKind::MUL {
+            circuit_builder.require_equal(
+                || "validate_prod_low_limb",
+                rs1_val * rs2_val,
+                prod_half.value() * (1u64 << BIT_WIDTH) + rd_val,
```

Comment:
You should be able to just shift the value directly:

```suggestion
                 (prod_half.value() << BIT_WIDTH) + rd_val,
```

(Please let me know, if the compiler complains.)

Reply:
We can't do that bcs it's an `Expression`

Reply:
See https://github.com/scroll-tech/ceno/pull/663